overlord/snapshotstate/backend/backend_test.go

```diff
@@ -324,26 +347,120 @@ func (s *snapshotSuite) TestIterReportsCloseError(c *check.C) {
 	c.Check(calledF, check.Equals, true)
 }
 
+func readerForFilename(fname string, c *check.C) *backend.Reader {
+	var snapname string
+	var id uint64
+	fn := strings.TrimSuffix(filepath.Base(fname), ".zip")
+	_, err := fmt.Sscanf(fn, "%d_%s", &id, &snapname)
+	c.Assert(err, check.IsNil, check.Commentf(fn))
+	f, err := os.Open(os.DevNull)
+	c.Assert(err, check.IsNil, check.Commentf(fn))
+	return &backend.Reader{
+		File: f,
+		Snapshot: client.Snapshot{
+			SetID: id,
+			Snap:  snapname,
+		},
+	}
+}
+
+func (s *snapshotSuite) TestIterIgnoresSnapshotsWithSetIdMismatches(c *check.C) {
+	logbuf, restore := logger.MockLogger()
+	defer restore()
+
+	defer backend.MockOsOpen(func(string) (*os.File, error) {
+		return new(os.File), nil
+	})()
+	readNames := 0
+	defer backend.MockDirNames(func(*os.File, int) ([]string, error) {
+		readNames++
+		if readNames > 2 {
```

Comment:
This should be "readNames > 1", the code in backend.go is a bit confusing, see my comment below about using a counter for "calledF". dirNames() will be called in a loop but all the relevant names are already returned in the first loop so no need to run it twice.

Reply:
Yes, `readNames > 1`, the code loops until it gets an error from dirNames(), so to mimic os.Readdirnames, we should return the list on the first call, and then io.EOF on the second.